primitives/trie/src/cache/mod.rs

```diff
@@ -0,0 +1,687 @@
+	/// The shared trie cache that created this instance.
+	shared: SharedTrieCache<H>,
+	/// The local cache for the trie nodes.
+	node_cache: Mutex<HashMap<H::Out, NodeOwned<H::Out>>>,
+	/// Keeps track of all the trie nodes accessed in the shared cache.
+	///
+	/// This will be used to ensure that these nodes are brought to the front of the lru when this
+	/// local instance is merged back to the shared cache.
+	shared_node_cache_access: Mutex<HashSet<H::Out>>,
+	/// The local cache for the values.
+	value_cache: Mutex<
+		HashMap<
+			ValueCacheKey<'static, H::Out>,
+			CachedValue<H::Out>,
+			BuildNoHashHasher<ValueCacheKey<'static, H::Out>>,
+		>,
+	>,
+	/// Keeps track of all values accessed in the shared cache.
+	///
+	/// This will be used to ensure that these nodes are brought to the front of the lru when this
+	/// local instance is merged back to the shared cache. This can actually lead to collision when
+	/// two [`ValueCacheKey`]s with different storage roots and keys map to the same hash. However,
+	/// as we only use this set to update the lru position it is fine, even if we bring the wrong
+	/// value to the top. The important part is that we always get the correct value from the value
+	/// cache for a given key.
+	shared_value_cache_access:
+		Mutex<HashSet<ValueCacheKeyHash, BuildNoHashHasher<ValueCacheKeyHash>>>,
+}
+
+impl<H: Hasher> LocalTrieCache<H> {
+	/// Return self as a [`TrieDB`](trie_db::TrieDB) compatible cache.
+	///
+	/// The given `storage_root` needs to be the storage root of the trie this cache is used for.
+	pub fn as_trie_db_cache<'a>(&'a self, storage_root: H::Out) -> TrieCache<'a, H> {
+		let value_cache = ValueCache::ForStorageRoot {
+			storage_root,
+			local_value_cache: self.value_cache.lock(),
+			shared_value_cache: self.shared.value_cache.read(),
+			shared_value_cache_access: self.shared_value_cache_access.lock(),
+		};
+
+		TrieCache {
+			shared_node_cache: self.shared.node_cache.read(),
+			local_cache: self.node_cache.lock(),
+			value_cache,
+			shared_node_cache_access: self.shared_node_cache_access.lock(),
+		}
+	}
+
+	/// Return self as [`TrieDBMut`](trie_db::TrieDBMut) compatible cache.
+	///
+	/// After finishing all operations with [`TrieDBMut`](trie_db::TrieDBMut) and having obtained
+	/// the new storage root, [`TrieCache::merge_into`] should be called to update this local
+	/// cache instance. If the function is not called, cached data is just thrown away and not
+	/// propagated to the shared cache. So, accessing these new items will be slower, but nothing
+	/// would break because of this.
+	pub fn as_trie_db_mut_cache<'a>(&'a self) -> TrieCache<'a, H> {
+		TrieCache {
+			shared_node_cache: self.shared.node_cache.read(),
+			local_cache: self.node_cache.lock(),
+			value_cache: ValueCache::Fresh(Default::default()),
+			shared_node_cache_access: self.shared_node_cache_access.lock(),
+		}
+	}
+}
+
+impl<H: Hasher> Drop for LocalTrieCache<H> {
+	fn drop(&mut self) {
```

Comment:
As far as I understand it is OK if updates to node cache and value cache are not in sync, but would be good to confirm that.

Reply:
Yes that is correct. Should I add a comment?